### PR TITLE
Update camhook to support new style camera detection

### LINKED ISSUE #90: How can I use official camera set on other motherboard
In GitLab by @superlaser177 on Oct 8, 2021, 01:00

Currently I run iidx in LDJ 120hz mode

And I have the official IIDX camera set run on new iidx pcb motherboard(c7b360)

But the game can not find the camera on any usb port,both two.

When I use camera hook in iidxhook.conf,only camera2 work but the image is from camera 1 with flickering(looks like some camera 2 image in it).

I have no idea about this problem,because both 2 camera have same hardware id,please help if you can,thanks!

### LINKED ISSUE #102: Camera hook doesn't work
In GitLab by @MatryoshkaLJY on Dec 2, 2022, 21:04

I played IIDX Casthour using the latest Bemanitools but it didn't detect my cameras. Here is part of the config file.
```
# Disable camera connection
io.disable_cams=false

# Disables the camera emulation
cam.disable_emu=true

# Override camera device ID detection (copy from device manager, do not escape)
cam.device_id1=USB\VID_1BCF&PID_2281&MI_00\6&1D224150&0&0000

# Override camera device ID detection (copy from device manager, do not escape)
cam.device_id2=USB\VID_046D&PID_0825&MI_00\6&2DD1CC96&0&0000
```
I'm not sure if the device ID I filled in is right. The information of the devices are in the screenshot below.
Untitled

### PR DESCRIPTION
**Pull request by willxinc**
This fixes #102 and resolves #90

Tested with a variety of cameras, working:
Logitech C922,
Logitech StreamCam

Not working
Logitech C615

### CODE REVIEW

**Comment by icex2** [comment]
> Not working
> Logitech C615

Any idea why it is not working? Would be nice if there is a way to test/tell if a cam is compatible.

File: src/main/camhook/cam-detect.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#ifndef CAMHOOK_CAM_DETECT_H
+#define CAMHOOK_CAM_DETECT_H
+
+#include <stdbool.h>
+#include <wchar.h>
+
+#define CAMERA_DATA_STRING_SIZE 0x100
+
+struct CameraData {
```
Comment (icex2) [inline src/main/camhook/cam-detect.h:10]: nit: code style `camera_data`

File: src/main/camhook/cam.h [hunk 2]
```diff
@@ -3,6 +3,13 @@
 
 #include "camhook/config-cam.h"
 
+enum camhook_version {
+    CAMHOOK_OLD,
+    CAMHOOK_NEW,
+};
```
Comment (icex2) [inline src/main/camhook/cam.h]: question: What's old vs. new and why keep the old "thing"?

File: dist/iidx/iidxhook-27.conf [hunk 3]
```diff
@@ -22,6 +22,9 @@ io.tt_multiplier=1.0
 # Disables the camera emulation
 cam.disable_emu=true
 
+# Camera port layout (0 = LDJ, 1 = CLDJ/TDJ-JA, 2 = TDJ-JB)
```
Comment (icex2) [inline dist/iidx/iidxhook-27.conf:25]: Can you explain what the port layout implies here regarding which one to chose, e.g. in a home setup? Might be something for the readme.

File: src/main/camhook/cam.c [hunk 4]
```diff
@@ -432,6 +504,180 @@ static HDEVINFO my_SetupDiGetClassDevsA(
...
+                        desc->idProduct,
+                        driverKeyNameA,
+                        camData[i].parent_driverKey);
+                    if (driverKeyNameA) {
+                        free(driverKeyNameA);
+                    }
+                    continue;
+                }
+
+                // TODO: check if driver key matches
```
Comment (icex2) [inline src/main/camhook/cam.c:608]: Still to be done?

File: src/main/camhook/cam.h (hunk 2 above)
Comment (willxinc) [reply src/main/camhook/cam.h]: it controls the vid/pid used, needs to be kept else old support breaks

File: src/main/camhook/cam.c (hunk 4 above)
Comment (willxinc) [reply src/main/camhook/cam.c:608]: ah forgot to remove the comment

**Comment by willxinc** [comment]
> > Not working
> > Logitech C615
> 
> Any idea why it is not working? Would be nice if there is a way to test/tell if a cam is compatible.

Not easily, I bought one to test, will look into it more in a future fix maybe, I suspect it's due to the supported resolution / features.

**Comment by icex2** [comment]
Overall, lgtm already regarding the code itself. A few documentation/information related things that I pointed out can be in a follow-up PR. Feel free to merge this once are done cleaning up.

**Comment by lutzchristoph** [comment]
Can’t get past the Camera check on sdvx-05.
Used a Sony ZV-E10 as a streaming device, id: “USB\VID_054C&PID_0DE3&MI_00\9&5B8827E&0&0000”
Is it possible to pass this test without a webcam?

**Comment by willxinc** [comment]
> Can’t get past the Camera check on sdvx-05. Used a Sony ZV-E10 as a streaming device, id: “USB\VID_054C&PID_0DE3&MI_00\9&5B8827E&0&0000” Is it possible to pass this test without a webcam?

The changes here are unrelated to SDVX / don't modify the existing camhook behavior, depending on the version it should just skip the camera check. For game-specific support like that I'm not sure this is the best forum for that.

**Pull request merged**